### ISSUE TITLE
Library re-organisation into pure squashfs libary and fuse implementation

### ISSUE THREAD

**Issue by ncw** [CalebQ42/squashfs#25]
I'm considering using this library in [rclone](https://github.com/rclone/rclone).

It works very well according to my tests.

However I find that it imports more stuff than I'd like ideally. Specifically to use the `squashfs` library as an archiver (like `archive/zip`) pulls in `github.com/CalebQ42/fuse` and `github.com/seaweedfs/fuse` which I don't need and will take up space in my binary.

It looks possible to re-arrange the library so it has an internal "pure" squashfs implementation then add the fuse implementation on top of that.

I think it could even be done in a backwards compatible way with some cunning use of type aliases, though it might be cleaner to declare a v2.

What do you think?

**Comment by CalebQ42** [comment]
tldr; It's possible, but would make the library unnecessarily confusing.

That was my initial plan for the FUSE implementation, but ultimately it's not really possible without making the library confusing. The way squashfs archives are set up are made for FUSE and so they contain a lot of information that doesn't make much sense in any other context. As an example, FUSE uses a file/folder's inode number to keep track of files and which file's belong to which folder and additionally needs the ability to convert an inode number to a file/folder. As an inode number is a fairly arbitrary number that has zero use outside of FUSE it doesn't make since to export it, especially since it's store within a Inode struct that also doesn't have any uses outside of a FUSE context.

Ultimately if you are creating an application where binary size is that important the best suggestion would be to create a copy of this library with the FUSE code stripped out, which should be fairly easy since all the FUSE code is separate.

**Comment by ncw** [comment]
> That was my initial plan for the FUSE implementation, but ultimately it's not really possible without making the library confusing. The way squashfs archives are set up are made for FUSE and so they contain a lot of information that doesn't make much sense in any other context. As an example, FUSE uses a file/folder's inode number to keep track of files and which file's belong to which folder and additionally needs the ability to convert an inode number to a file/folder. As an inode number is a fairly arbitrary number that has zero use outside of FUSE it doesn't make since to export it, especially since it's store within a Inode struct that also doesn't have any uses outside of a FUSE context.

I noticed this when I was fiddling with the code that the fuse methods look at unexported internals.

I would disagree that this will make the library more confusing though. Perhaps someone has a good use for that inode number that we don't know about.

For example [`archive/zip` exports](https://pkg.go.dev/archive/zip#FileHeader) all the information about the zip archive it can, including somewhat esoteric things like the CRC32 of the data and the extra fields it doesn't know how to interpret.

> Ultimately if you are creating an application where binary size is that important the best suggestion would be to create a copy of this library with the FUSE code stripped out, which should be fairly easy since all the FUSE code is separate.

That would indeed be easy - it just seems a shame to have to fork the library to do this.

I'd be willing to contribute a PR to split the library like this if you are interested, but I understand if you'd like to leave the library as-is.

Thank you :-)

**Comment by CalebQ42** [comment]
I'll consider it. If you want to put the work into making a PR, I'll definitely look at it, but no promises if it'll get merged or not.

**Comment by CalebQ42** [comment]
Just wanted to let you know I decided to work on this (<https://github.com/CalebQ42/squashfs/tree/exp1>). `squashfs/squashfs` will be the "low" level library that has a lot more information exposed and then have the main library that's basically just a wrapper around it that is friendlier to use.

**Comment by CalebQ42** [comment]
Implemented as of `v1.0`
